### ISSUE TITLE
Nested mapped type fails recognizing excess property

### ISSUE THREAD

**Issue by benjamin-hg** [Microsoft/TypeScript#17482]
**TypeScript Version:**  2.4.0 

**Code**

```ts

export type MyNestedMappedType<TModel> = {
    [P in keyof TModel]: string | {
        [Q in keyof TModel[P]]: string;
    };
}

declare function For<TModel>(): {
    useMapped<TMapped extends MyNestedMappedType<TModel>>(mappedModel: TMapped);
};


export interface SomeNestedModel {
    innerProp: string,
}

export interface SomeModel {
    outerProp: string;
    nested: SomeNestedModel;
}


For<SomeModel>().useMapped({
    outerProp: 'some value',
    nested: {
        innerProp: 'some value', // <-- this is strange: when commenting out this line, the line below (num2: ...) gets marked as an error (which is desired).
        innerProp2: 'some value',
    },
    num2: 'some value', // no error! (but would expect one here)
});

For<SomeModel>().useMapped({
    outerProp: 'some value',
    nested: {
        // innerProp: 'some value', // <-- this is strange: when commenting out this line, the line below (num2: ...) gets marked as an error (which is desired).
        innerProp2: 'some value',
    },
    num2: 'some value', // error: Object literal may only specify known properties, and 'num2' does not exist in type 'MyNestedMappedType<SomeModel>'
});
```

**Expected behavior:**
` num2: 'some value'` should cause an error in both cases

**Actual behavior:**
` num2: 'some value'` causes an error only in the latter case.

(If this is by design, I'd be happy if someone could explain.)

**Comment by mhegazy** [comment]
In the first example the object literal in question is used to infer a generic type parameter for the call of useMapped, so the error is not expected.

in the second case, missing `innerProp` cuases the constraint on `useMapped` to fail. the error reporting here is the same as assigining to a vlue of the same type as the constrinat, i.e.:

```ts
x = {
    outerProp: 'some value',
    nested: {
         // innerProp: 'some value', // <-- this is strange: when commenting out this line, the line below (num2: ...) gets marked as an error (which is desired).
        innerProp2: 'some value',
    },
    num2: 'some value', // error: Object literal may only specify known properties, and 'num2' does not exist in type 'MyNestedMappedType<SomeModel>'
}
```

The error on `num2` happens to be reported earlier than the other one, and that what you see.

**Comment by benjamin-hg** [comment]
Hi @mhegazy, 
first, thanks for explaining. I understood the following:
 - if the assigned value has a 'derived' type, meaning all properties of the super type (`SomeModel`) are in place (and maybe others), the assignment is valid.
 - if it is not a derived type, it has to be at lease the *same* type, which fails in this case with `Object literal may only specify known properties, and 'num2' does not exist in type 'MyNestedMappedType<SomeModel>'`

If I got this right, the error message is still confusing, because it contradicts the validity of a derived type (implying unknown properties are *actually allowed*)
So my suggestion would be to fix the error message in some way.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
